### PR TITLE
Chore/react testing

### PR DESCRIPTION
**Pull request by kacper-echaust**

### CODE REVIEW

File: src/tests/NewCard.test.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import { CardSide, NewCard } from '../components/AppMain/NewCard/NewCard'
+import { Buttons } from '../components/AppMain/NewCard/Buttons/Buttons'
+import { AppHeader } from '../components/AppHeader'
+
+test('error alert when front or back side is empty', () => {
+	const onNextMock = jest.fn()
+	const onSaveMock = jest.fn()
+    render(<AppHeader/>)
+	render(<NewCard />)
+	render(
```
Comment (Sumick) [inline src/tests/NewCard.test.tsx]: **issue:** You should not render these components separately as you don't test integration this way. You should render whole parent that has these components included

File: src/tests/NewCard.test.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		<Buttons side={CardSide.Front} onCancel={() => {}} onSave={onSaveMock} onBack={() => {}} onNext={() => {}} />
+	)
+	const saveButton = screen.getByRole('button', { name: /Save/i })
+	fireEvent.change(input, {
+		target: { value: '' },
+	})
+
+	fireEvent.click(saveButton)
+
+	expect(alertSpy).toHaveBeenCalledWith('Please enter front and back card.')
```
Comment (Sumick) [inline src/tests/NewCard.test.tsx]: **issue:** you should check if save button is disabled or any error is displayed.

File: src/tests/NewCard.test.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	)
+	const saveButton = screen.getByRole('button', { name: /Save/i })
+    
+	fireEvent.change(input, {
+		target: { value: 'Back Side' },
+	})
+
+	fireEvent.click(saveButton)
+
+	expect(alertSpy).toHaveBeenCalledWith('Form successfully submitted!')
```
Comment (Sumick) [inline src/tests/NewCard.test.tsx]: **issue:** This does not check if card was added. You should render whole app, add new card and then check if new card was added to your list

**Review by Sumick — CHANGES_REQUESTED** [review]
These test are not integrations tests - they are more like unit tests where you just test if elements were clicked. To test integration, you should render whole group of components at once - then use `screen` method to access and simulate user interactions. Check this great article from a co-creator of react-testing-library for amazing tips: https://kentcdodds.com/blog/common-mistakes-with-react-testing-library

File: src/tests/NewCard.test.tsx [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import React from 'react'
+import { fireEvent, render, screen } from '@testing-library/react'
+import { App } from '../App.tsx'
+
+test('error alert when front or back side is empty', () => {
```
Comment (Sumick) [inline src/tests/NewCard.test.tsx]: **nit:** This test tests if card was added, not if error is present

**Review by Sumick — CHANGES_REQUESTED** [review]
Yup, you got into good direction! That's an integration test now, good job :) Now you need to add second test for checking this first scenario which is: "It should not be possible to add a flashcard when front or back card value is empty"

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
## Code Review Report

### Task Goal
Write an **integration test** covering adding flashcard.

### Requirements
1. It should not be possible to add a flashcard when front or back card value is empty.
2. It should be possible to add a flashcard when front, and back values are present.

### Code Review

#### package.json
The changes in `package.json` include the addition of dependencies and scripts necessary for testing:
- Added `react-scripts` for running tests.
- Added testing libraries: `@testing-library/jest-dom`, `@testing-library/react`, `@testing-library/user-event`.
- Added a test script: `"test": "react-scripts test"`.

These changes are appropriate and necessary for writing and running integration tests.

#### src/components/AppMain/NewCard/Buttons/Buttons.tsx
The `Buttons` component has been modified to include a `disabled` prop. However, this prop is not being used within the component to disable the buttons based on the card values.

##### Suggested Improvement:
To meet the requirement that a flashcard cannot be added when either the front or back value is empty, the `disabled` prop should be applied to the "Save" button.

```jsx
<button className={styles.nextButton} {...propsSecondButton} disabled={propsSecondButton.value === 'Save' && props.disabled}>
    {propsSecondButton.value}
</button>
```

#### src/components/AppMain/NewCard/NewCard.tsx
The `NewCard` component has been updated to include logic that prevents saving a card if either the front or back value is empty.

- The `handleSave` function now checks if both `card.front` and `card.back` are non-empty before proceeding.
- The `Buttons` component is passed a `disabled` prop based on whether either `card.front` or `card.back` is empty.

These changes align with the requirement that a flashcard cannot be added when either the front or back value is empty.

#### src/tests/NewCard.test.tsx
Two tests have been added:
1. **add card**: This test covers adding a flashcard when both front and back values are present.
2. **can not add empty card**: This test attempts to add a flashcard with empty values and ensures that it is not possible.

##### Bugs:
- In the **can not add empty card** test, the assertion `expect(saveButton).toBeFalsy` is incorrect. It should check if the button is disabled instead.

##### Suggested Improvement:
Update the test to check if the "Save" button is disabled when either input is empty.

```jsx
test('can not add empty card', () => {
    render(<NewCard />)

    const input = screen.getByRole('textbox') as HTMLInputElement
    const nextButton = screen.getByRole('button', { name: /next/i })

    expect(input.value).toBe('')

    fireEvent.click(nextButton)

    const saveButton = screen.getByRole('button', { name: /save/i })

    expect(input.value).toBe('')
    
    expect(saveButton).toBeDisabled()
})
```

### Summary
The code changes mostly cover the task requirements. However, there are some improvements needed:
1. The `disabled` prop should be applied to the "Save" button in the `Buttons` component.
2. The test for ensuring a flashcard cannot be added when values are empty should check if the "Save" button is disabled.

With these improvements, the code will fully meet the task requirements.

**Review by NerdbordMentor — APPROVED** [review]
## Code Review Report

### Task Goal
Write an **integration test** covering adding flashcard.

### Requirements
1. It should not be possible to add a flashcard when front or back card value is empty.
2. It should be possible to add a flashcard when front, and back values are present.

### Review of Code Changes

#### package.json
The changes in `package.json` include the addition of testing libraries and scripts necessary for running tests:
- Added `react-scripts` and testing libraries such as `@testing-library/react`, `@testing-library/jest-dom`, and `@testing-library/user-event`.
- Added a `test` script to run tests using `react-scripts`.

These changes are appropriate for enabling the testing environment required for the task.

#### src/components/AppMain/NewCard/Buttons/Buttons.tsx
The changes in `Buttons.tsx` include:
- Addition of a `disabled` prop to the save button.
- Conditional disabling of the save button based on whether the front or back card values are empty.

These changes ensure that the save button is disabled when either the front or back card value is empty, which aligns with the first requirement.

#### src/components/AppMain/NewCard/NewCard.tsx
The changes in `NewCard.tsx` include:
- Modification of the `handleSave` function to check if both `card.front` and `card.back` are present before saving the card.
- Passing a `disabled` prop to the `Buttons` component to disable the save button when either `card.front` or `card.back` is empty.

These changes ensure that a flashcard cannot be added if either the front or back value is empty, and allow adding a flashcard only when both values are present, which aligns with both requirements.

#### src/tests/NewCard.test.tsx
The new test file `NewCard.test.tsx` includes two tests:
1. **add card**: This test covers adding a flashcard when both front and back values are present.
   - It simulates user interactions to add a new card with both front and back values.
   - It verifies that the card is added successfully by checking for the presence of the front text.

2. **can not add empty card**: This test covers the scenario where it should not be possible to add a flashcard when either front or back value is empty.
   - It simulates user interactions to attempt adding a card with empty values.
   - It verifies that the save button is disabled when either value is empty.

These tests cover both requirements specified in the task.

### Summary
The provided code changes cover the task requirements effectively:
1. The save button is disabled when either the front or back card value is empty, preventing the addition of an incomplete flashcard.
2. The integration tests verify that a flashcard can only be added when both front and back values are present, and cannot be added when either value is empty.

No bugs were identified in the provided code changes. The implementation meets the task requirements comprehensively.
